### ISSUE TITLE
Virtual Network expects all subnets to be defined inline, causes deletion-recreation churn otherwise

### ISSUE THREAD

**Issue by Josh-Tilles** [pulumi/pulumi-azure-native#611]
A repository demonstrating the following can be found at https://github.com/MerelyAPseudonym/subnet-churn-repro.

I expected the familiar all-or-nothing constraint regarding inline subnet definitions to carry over to the nextgen provider, but even if I _only_ use standalone subnet resources I'm observing unstable behavior. The initial run works, but then **without changing the code** subsequent runs delete and recreate the subnet.

Initialize both network resources:
```sh-session
$ pulumi up --refresh --non-interactive --yes --skip-preview --suppress-permalink
Updating (dev)

 ~  pulumi:pulumi:Stack subnet-churn-repro-dev refreshing 
 ~  pulumi:pulumi:Stack subnet-churn-repro-dev refreshing 
 ~  azure-nextgen:resources/v20200601:ResourceGroup resourceGroup refreshing 
 ~  azure-nextgen:resources/v20200601:ResourceGroup resourceGroup refresh 
    pulumi:pulumi:Stack subnet-churn-repro-dev running 
    azure-nextgen:resources/v20200601:ResourceGroup resourceGroup  
 +  azure-nextgen:network/v20200601:VirtualNetwork network creating 
 +  azure-nextgen:network/v20200601:VirtualNetwork network created 
 +  azure-nextgen:network/v20200601:Subnet subnet creating 
 +  azure-nextgen:network/v20200601:Subnet subnet created 
@ Updating....
    pulumi:pulumi:Stack subnet-churn-repro-dev  
 
Resources:
    + 2 created
    2 unchanged
```

Double-check that the subnet exists via the Azure CLI:
```sh-session
$ az network vnet subnet show --resource-group=josh-repro --vnet-name=net --name=net-sub1
{
  "addressPrefix": "10.0.0.0/24",
…
```

Rerun Pulumi without changing the code:
```sh-session
$ pulumi up --refresh --non-interactive --yes --skip-preview --suppress-permalink
Updating (dev)

 ~  pulumi:pulumi:Stack subnet-churn-repro-dev refreshing 
 ~  pulumi:pulumi:Stack subnet-churn-repro-dev refreshing 
 ~  azure-nextgen:network/v20200601:Subnet subnet refreshing 
 ~  azure-nextgen:resources/v20200601:ResourceGroup resourceGroup refreshing 
 ~  azure-nextgen:network/v20200601:VirtualNetwork network refreshing 
 ~  azure-nextgen:resources/v20200601:ResourceGroup resourceGroup refresh 
 ~  azure-nextgen:network/v20200601:VirtualNetwork network refresh [diff: ~__inputs,etag,subnets]
 ~  azure-nextgen:network/v20200601:Subnet subnet refresh 
    pulumi:pulumi:Stack subnet-churn-repro-dev running 
    azure-nextgen:resources/v20200601:ResourceGroup resourceGroup  
 ~  azure-nextgen:network/v20200601:VirtualNetwork network updating [diff: -subnets]
 ~  azure-nextgen:network/v20200601:VirtualNetwork network updated [diff: -subnets]
    azure-nextgen:network/v20200601:Subnet subnet  
    pulumi:pulumi:Stack subnet-churn-repro-dev  
 
Resources:
    ~ 1 updated
    3 unchanged
```

Run the same Azure CLI command, finding the subnet absent:
```sh-session
$ az network vnet subnet show --resource-group=josh-repro --vnet-name=net --name=net-sub1
ResourceNotFoundError: Resource /subscriptions/123e4567-e89b-12d3-a456-426614174000/resourceGroups/josh-repro/providers/Microsoft.Network/virtualNetworks/net/subnets/net-sub1 not found.
```

Again, without changing the code, rerun Pulumi:
```sh-session
$ pulumi up --refresh --non-interactive --yes --skip-preview --suppress-permalink
Updating (dev)

 ~  pulumi:pulumi:Stack subnet-churn-repro-dev refreshing 
 ~  pulumi:pulumi:Stack subnet-churn-repro-dev refreshing 
 ~  azure-nextgen:resources/v20200601:ResourceGroup resourceGroup refreshing 
 ~  azure-nextgen:network/v20200601:Subnet subnet refreshing 
 ~  azure-nextgen:network/v20200601:VirtualNetwork network refreshing 
 ~  azure-nextgen:resources/v20200601:ResourceGroup resourceGroup refresh 
 ~  azure-nextgen:network/v20200601:Subnet subnet refresh 
 ~  azure-nextgen:network/v20200601:VirtualNetwork network refresh 
    pulumi:pulumi:Stack subnet-churn-repro-dev running 
    azure-nextgen:resources/v20200601:ResourceGroup resourceGroup  
    azure-nextgen:network/v20200601:VirtualNetwork network  
 +  azure-nextgen:network/v20200601:Subnet subnet creating 
 +  azure-nextgen:network/v20200601:Subnet subnet created 
    pulumi:pulumi:Stack subnet-churn-repro-dev  
 
Resources:
    + 1 created
    3 unchanged
```

It just goes back and forth.

**Comment by keithharvey** [comment]
We hit this today and luckily were still about 6 hours from being in production with it so rebuilding the cluster wasn't an issue. I'm going to try using the azure provider for the vnet definition and the azure-native provider for the subnet definition, hopefully that prevents the issue from cropping up again but I will report back if it does.

**Comment by andrekiba** [comment]
Hi, any news about this?

**Comment by pwinstead** [comment]
Add my vote against solution 2. Obviously, it's based on my use case but maybe others have similar reason for wanting to manage subnets independent from the vnet.
In my scenario I have a Vnet being created in a 'foundation' Pulumi project and ignore the subnet property. This project builds out resources that are widely shared.  Then I have other Pulumi projects that build up additional Azure resources specific to their needs. These project could be adding subnets to the existing vnet.

**Comment by joeduffy** [comment]
@danielrbradley What does ARM/Bicep do? My inclination would be to follow in their footsteps here.

**Comment by mikhailshilkov** [comment]
@joeduffy AFAIK, we already do the same as ARM/Bicep: they have the same problem.

This issue has been open for seven years: https://github.com/Azure/azure-quickstart-templates/issues/2786

A Bicep issue closed with the link to the one above: https://github.com/Azure/bicep/issues/5397

It sounds like they vaguely plan to solve it on the service side: https://github.com/Azure/azure-quickstart-templates/issues/2786#issuecomment-1251517143

**Comment by danielrbradley** [comment]
Notes from our internal ideation:

Something similar to option (3) seems most feasible, but adopting an approach similar to Azure's comment - ignoring the subnets property if not specified. 

- Keep the subnets property, but add a deprecation notice explaining the potential issues of conflicting with externally specified subnets. 
- Include in the description explanation that if this property is left empty then the subnets will remain unchanged.
- In the provider metadata, indicate that this property should be ignored if not specified.
- In the provider implementation, where the metadata indicates that a property should be ignored if empty and the property is empty, perform a read of the resource before the update then copy the property value into the update body.
- Plan to remove this property at the next major version of the provider to avoid possible ambiguity of how to use the resource. The migration path would require anyone using inline subnets to re-import the subnets as stand-alone resources.

One edge case is that if anyone previously using the subnets property goes to do an update to remove all subnets from the VNet then the subnets will not be removed.

Option (4) is still possible but is likely more complicated an implementation to get correct and so we'd prefer leaning toward the simplicity of (3). To elaborate on the discussed approach: 
- Keep the SDK exactly the same - with both the inline and stand-alone resources.
- Indicate in the metadata that this property also exists as a stand-alone resource.
- If a resource has properties which also exist as stand-alone resources, then we:
  - Calculate a diff of the subnets property between the old state and the new inputs (which have been removed, added or updated)
  - Perform a read before the update.
  - Apply the diff (additions, removals and updates) to the subnets which were read.
  - Use the updated subnets in the body of the update.

Both these approaches should be implemented in a generic manner (using metadata rather than hard-coded exceptions) so that it's not specific to this single resource as there's other resources which similar hierarchical structures.

**Comment by danielrbradley** [comment]
The fix for this issue is available in [v2.8.0 of the azure-native provider](https://github.com/pulumi/pulumi-azure-native/releases/tag/v2.8.0).

**Comment by rraval** [comment]
@danielrbradley I've been following this issue since I had assumed that it was related to #579, where `containerservice:ManagedCluster.agentPoolProfiles` has a similar parent-child relationship.

However, looking over the fix (#2755), I find no mention of `ManagedCluster`. Are these two separate issues then?

**Comment by danielrbradley** [comment]
Hi @rraval thanks for reaching out. I've just had a read through #579 and though it is somewhat related to nested resources, the behaviour issue is very different.

It is interesting that it doesn't get automatically matched as a sub-resource property. On investigating, this is because the sub-property is not exactly the same shape as the standalone resource. 
- The standalone resource references `#/definitions/AgentPool`
- The sub-property references `#/definitions/ManagedClusterAgentPoolProfile`
- Both of the above then extend `#/definitions/ManagedClusterAgentPoolProfileProperties`
- `#/definitions/ManagedClusterAgentPoolProfile` also then adds an extra `name` property.

This would be a much more complex scenario to match, and actually based on the behaviour I think we actually wouldn't want to implement the same fix as for this issue.

I'll detail a bit more of my finding of the behaviour on #579
